crates/subspace-farmer/src/piece_cache.rs

```diff
@@ -230,35 +232,74 @@ where
-                stored_pieces.clear();
-                stored_pieces.reserve(contents.len());
-                free_offsets.clear();
-
-                for (offset, maybe_piece_index) in contents {
-                    match maybe_piece_index {
-                        Some(piece_index) => {
-                            stored_pieces
-                                .insert(RecordKey::from(piece_index.to_multihash()), offset);
-                        }
-                        None => {
-                            free_offsets.push(offset);
-                        }
-                    }
+            .enumerate()
+            .map(
+                |(index, ((mut stored_pieces, mut free_offsets), new_cache))| {
+                    run_future_in_dedicated_thread(
+                        move || async {
+                            let contents = new_cache.contents();
+                            stored_pieces.reserve(contents.len());
+
+                            for (offset, maybe_piece_index) in contents {
+                                match maybe_piece_index {
+                                    Some(piece_index) => {
+                                        stored_pieces.insert(
+                                            RecordKey::from(piece_index.to_multihash()),
+                                            offset,
+                                        );
+                                    }
+                                    None => {
+                                        free_offsets.push(offset);
+                                    }
+                                }
+
+                                // Allow for task to be aborted
+                                yield_now().await;
+                            }
+
+                            DiskPieceCacheState {
+                                stored_pieces,
+                                free_offsets,
+                                backend: new_cache,
+                            }
+                        },
+                        format!("piece-cache.{index}"),
+                    )
+                },
+            )
+            .collect::<Result<Vec<_>, _>>();
+
+        let caches_futures = match maybe_caches_futures {
+            Ok(caches_futures) => caches_futures,
+            Err(error) => {
+                error!(
+                    %error,
+                    "Failed to spawn piece cache reading thread"
+                );
+
+                return;
+            }
+        };
+
+        let mut caches = Vec::with_capacity(caches_futures.len());
+        let mut caches_futures = caches_futures.into_iter().collect::<FuturesOrdered<_>>();
+
+        while let Some(maybe_cache) = caches_futures.next().await {
+            match maybe_cache {
+                Ok(cache) => {
+                    caches.push(cache);
                 }
+                Err(_cancelled) => {
+                    error!("Piece cache reading thread panicked");
```

Comment:
Why don't include an error in the message? Is it a generic "Task aborted" string?

Reply:
There is no error there, it is just "Cancelled" message that will be printed there. That is due to API of `run_future_in_dedicated_thread`.